Cannot using `sending` with `some` return value

### Description

Hmm, shouldn't this be allowed? It is a result!

```
test.swift:7:21: error: 'sending' may only be used on parameters and results
 5 | }
 6 | 
 7 | func makeThing() -> sending some MyProtocol {
   |                     `- error: 'sending' may only be used on parameters and results
 8 |     NonSendable()
 9 | }
```

### Reproduction

```swift
protocol MyProtocol {
}

class NonSendable: MyProtocol {
}

func makeThing() -> sending some MyProtocol {
    NonSendable()
}
```


### Expected behavior

No errors

### Environment

Apple Swift version 6.0-dev (LLVM 5b37bd4c1c578e9, Swift 54898ce105c7185)
Target: arm64-apple-macosx14.0

### Additional information

_No response_

--- a/lib/Sema/TypeCheckGeneric.cpp
+++ b/lib/Sema/TypeCheckGeneric.cpp
@@ -225,10 +225,17 @@ OpaqueResultTypeRequest::evaluate(Evaluator &evaluator,
     opaqueDecl->setGenericSignature(GenericSignature());
   }
 
+  TypeResolverContext resolverContext = TypeResolverContext::FunctionResult;
+  if (isa<VarDecl>(originatingDecl)) {
+    // Non-opaque result types are resolved against this context, so
+    // replicate that logic here.
+    resolverContext = TypeResolverContext::PatternBindingDecl;
+  }
+
   // Resolving in the context of `opaqueDecl` allows type resolution to create
   // opaque archetypes where needed
   auto interfaceType =
-      TypeResolution::forInterface(opaqueDecl, TypeResolverContext::None,
+      TypeResolution::forInterface(opaqueDecl, resolverContext,
                                    /*unboundTyOpener*/ nullptr,
                                    /*placeholderHandler*/ nullptr,
                                    /*packElementOpener*/ nullptr)

--- a/test/Concurrency/sending_subscript.swift
+++ b/test/Concurrency/sending_subscript.swift
@@ -2,7 +2,9 @@
 
 // REQUIRES: concurrency
 
-class NonSendableKlass {}
+protocol P {}
+
+class NonSendableKlass: P {}
 
 // CHECK-DAG: subscript(_: sending NonSendableKlass) -> sending NonSendableKlass { get }
 
@@ -20,3 +22,18 @@ struct S2 {
   }
 }
 
+// CHECK-DAG: subscript(_: sending NonSendableKlass) -> sending some P { get }
+
+// CHECK-DAG: sil hidden [ossa] @$s17sending_subscript2S3VyQrAA16NonSendableKlassCncig : $@convention(method) (@sil_sending @owned NonSendableKlass, S3) -> @sil_sending @out @_opaqueReturnTypeOf("$s17sending_subscript2S3VyQrAA16NonSendableKlassCncip", 0) __ {
+struct S3 {
+  subscript(_: sending NonSendableKlass) -> sending (some P) { NonSendableKlass() }
+}
+
+// CHECK-DAG: subscript(_: sending NonSendableKlass) -> sending (some P, NonSendableKlass) { get }
+
+// CHECK-DAG: sil hidden [ossa] @$s17sending_subscript2S4VyQr_AA16NonSendableKlassCtAEncig : $@convention(method) (@sil_sending @owned NonSendableKlass, S4) -> (@sil_sending @out @_opaqueReturnTypeOf("$s17sending_subscript2S4VyQr_AA16NonSendableKlassCtAEncip", 0) __, @sil_sending @owned NonSendableKlass) {
+struct S4 {
+  subscript(_: sending NonSendableKlass) -> sending (some P, NonSendableKlass) {
+    (NonSendableKlass(), NonSendableKlass())
+  }
+}

--- a/test/Concurrency/transfernonsendable_sending_results.swift
+++ b/test/Concurrency/transfernonsendable_sending_results.swift
@@ -41,9 +41,11 @@ struct CustomActor {
 @CustomActor func transferToCustomDirect(_ t: NonSendableKlass) async {}
 func useValueIndirect<T>(_ t: T) {}
 func useValueDirect(_ t: NonSendableKlass) {}
+func useValueOpaque(_ t: some OpaqueP) {}
 
 func transferValueDirect(_ x: sending NonSendableKlass) {}
 func transferValueIndirect<T>(_ x: sending T) {}
+func transferValueOpaque(_ t: sending some OpaqueP) {}
 
 func transferResult() -> sending NonSendableKlass { NonSendableKlass() }
 func transferResultWithArg(_ x: NonSendableKlass) -> sending NonSendableKlass { NonSendableKlass() }
@@ -95,6 +97,12 @@ func sendParameter<T>(_ t: sending T) {}
 actor MyActor {
   private var ns = NonSendableKlass()
 }
+
+protocol OpaqueP {}
+extension NonSendableKlass: OpaqueP {}
+
+func opaqueSendingResult() -> sending (some OpaqueP) { NonSendableKlass() }
+func opaqueSendingResultAsync() async -> sending (some OpaqueP) { NonSendableKlass() }
 
 /////////////////
 // MARK: Tests //
@@ -319,4 +327,42 @@ extension MyActor {
 
     return value
   }
+}
+
+////////////////////////////////
+// MARK: Opaque Results Tests //
+////////////////////////////////
+
+func test_opaque_sending_result_basic() {
+ let x = opaqueSendingResult()
+ let y = opaqueSendingResult()
+
+ sendParameter(x)
+ sendParameter(y)
+}
+
+func test_opaque_sending_result_async() async {
+  let x = await opaqueSendingResultAsync()
+  useValueIndirect(x)
+}
+
+@MainActor
+func test_opaque_sending_result_main_actor() async {
+  let x: some OpaqueP = await opaqueSendingResultAsync()
+  sendParameter(x)
+  // expected-warning @-1 {{sending 'x' risks causing data races}}
+  // expected-note @-2 {{'x' used after being passed as a 'sending' parameter; Later uses could race}}
+  useValueOpaque(x)
+  // expected-note @-1 {{access can happen concurrently}}
+}
+
+func test_opaque_sending_result_async_race_detected() async {
+  let x = await opaqueSendingResultAsync()
+  useValueIndirect(x)
+  sendParameter(x)
+}
+
+struct TestOpaqueSendingSubscript {
+  subscript(i: Int) -> sending some OpaqueP { NonSendableKlass() }
+  subscript(send i: sending Int) -> sending some OpaqueP { NonSendableKlass() }
 }

--- a/test/Parse/sending.swift
+++ b/test/Parse/sending.swift
@@ -13,7 +13,18 @@ func testArgResult(_ x: sending String) -> sending String {
 }
 
 func testVarDeclDoesntWork() {
-  var x: sending String // expected-error {{'sending' may only be used on parameter}}
+  var x: sending String // expected-error {{'sending' may only be used on parameters and results}}
+  let y: sending String // expected-error {{'sending' may only be used on parameters and results}}
+
+  struct S {
+    var opaqueStored: sending some Equatable = 0 // expected-error {{'sending' may only be used on parameters and results}}
+    let opaqueStoredImmutable: sending some Equatable = 0 // expected-error {{'sending' may only be used on parameters and results}}
+    var opaqueComputed: sending some Equatable { 0 } // expected-error {{'sending' may only be used on parameters and results}}
+    var opaqueWriteback: sending some Equatable { // expected-error {{'sending' may only be used on parameters and results}}
+      get { 0 }
+      set {}
+    }
+  }
 }
 
 func testVarDeclTupleElt() -> (sending String, String) {} // expected-error {{'sending' cannot be applied to tuple elements}}

--- a/test/Sema/diag_erroneous_iuo.swift
+++ b/test/Sema/diag_erroneous_iuo.swift
@@ -323,9 +323,6 @@ func returnsFunc2Identifier() -> (Int) -> ImplicitlyUnwrappedOptional<Int> { // 
 }
 
 func opaqueResult() -> (some Equatable)! { return 1 }
-// expected-swift4-warning@-1:40 {{using '!' here is deprecated; this is an error in the Swift 5 language mode}}{{none}}
-// expected-swift5-error@-2:40 {{'!' is not allowed here}}{{none}}
-// expected-note@-3:40 {{use '?' instead}}{{40-41=?}}
 
 let x0 = 1 as ImplicitlyUnwrappedOptional // expected-error {{'ImplicitlyUnwrappedOptional' has been renamed to 'Optional'}}{{15-42=Optional}}
 

--- a/test/Sema/sending.swift
+++ b/test/Sema/sending.swift
@@ -22,3 +22,8 @@ func test_repeated_sending_mixed(_ x: sending consuming sending inout Int) {}
 // Just until we get the results setup.
 func test_sending_result_in_tuple() -> (sending Int, Int) {}
 // expected-error @-1 {{'sending' cannot be applied to tuple elements}}
+
+// https://github.com/swiftlang/swift/issues/74846
+
+func test_good_returns_sending_opaque() -> sending (some Equatable) { 0 }
+func test_good_returns_sending_opaque_in_tuple() -> sending (some Equatable, Int) { (0, 0) }

